### DISCUSSION TITLE
Fast method for updating operating point

### DISCUSSION THREAD

**Discussion by Hofsmo** [NREL-SIIP/PowerSimulationsDynamics.jl#225] [Q&A]
Hi,
I am currently working on small signal analysis and I want to see how the eigenvalues change with some parameters. For instance I want to see how the eigenvalues change if I change the active power of a generator. From the documentation I see that I can do `jacobian(JM, x)`, however, I am not sure what the best method for finding x in the new operating state is. I tried two different methods. One is to create a new simulation struct after I changed the power and get the new x from the simulation struct. The other method was to change your code to export the function `set_operating_point!`. The second method is a lot faster, so I was wondering if it would make sense to let this method be available to the user by default. I was also wondering if you already have implemented a smart way of doing this.

Cheers Sigurd

**Comment by Hofsmo** [comment]
If you are interested I can share my changes. It was not much that had to be changed. At least I found it very usefult to be able to reinitialise the system after changing things like active power set point without rebuilding the whole system.

**Reply by rodrigomha** [reply]
This could be useful.

Most of the time, since I need to reinitialize the system by running a new power flow, I tend to re-do a Simulation, The main concern that I have if that if I change some parameter, it is likely that the system will not be in steady state, so a Eigenvalue analysis may not be the best in that case.

In any case, this is super dependent of what you are trying, but I would like to see what was your approach for that. If you want you can do a pull request from a fork and we can take a look with @jd-lara too see if it would be useful to have that in master.

**Reply by Hofsmo** [reply]
Hi,
Many thanks for the nice answer.

I am looking into eigenvalue sensitivity to generator set points like how they do it in this paper https://www.sciencedirect.com/science/article/pii/S0378779618300531 . I know I can calculate these senstivities analytically, but I need to familiarise myself more with your library before I manage to do it analytically. In any case I like to be able to do it numerically too, since it works as a nice base case.

Regarding the steady state concern. I assumed that the various initialize methods in the method set_operating_point! method calculated the steady state, however, I see now that the Simulation constructor does some more things to the initial state. In any case I get the same result using my method and when I redo a Simulation. My idea was that if I get the jacobian functor once, I could make small changes to and then get the new initial_conditions using set_operating_point! and use those with the functor. In this way I would not need to recreate the system and recalculate the jacobian functor in every iteration. 

After reading your reply I got a bit suspcious that I may have misunderstood how things are implemented, so I want to check if my solution works on more cases.  If it does I can make a small example script and a PR, otherwise I have learned more about your nice library.

**Reply by jd-lara** [reply]
The constructor of the simulation tries to ensure that the models don't deviate from the power flow solution and that the solution of the non-linear system for the same model is in equilibrium. Since we want to support new models from others, we needed a routine agnostic to new models that we can't anticipate. 

I have a working PR with improvements to the simulation speed and avoid remaking all the initialization but haven't had a chance to finish it.

**Reply by Hofsmo** [reply]
I see that makes sense. I found the PR now, I will check it out. Many thanks for the good answers :)

**Reply by jd-lara** [reply]
@Hofsmo thanks for the interest, this is an area of improvement we need to work on.

**Comment by jd-lara** [comment]
Sorry we missed this, please feel free to open a PR anytime. We are working on a similar solution but help is welcome.
